source/MaterialXFormat/GraphIO.h

```diff
@@ -0,0 +1,436 @@
+    /// The default orientation is set to TOP_DOWN which means the root is
+    /// at the top and upstream downs are positioned downward from the root.
+    /// @param val Orientationn hint
+    void setOrientation(Orientation val)
+    {
+        _orientation = val;
+    }
+
+    /// Get whether option to write subgraphs is enabled.
+    /// @return Orientation hint
+    Orientation getOrientation() const
+    {
+        return _orientation;
+    }
+
+  protected:
+    /// Write category labels
+    bool _writeCategories = true;
+
+    /// Write subgraphs
+    bool _writeSubgraphs = true;
+
+    /// Orientation hint
+    Orientation _orientation = Orientation::TOP_DOWN;
+};
+
+/// @class GraphIO
+/// <summary>
+///     Interface defining classes which interpret a GraphElement and 
+///     generate output in the desired output format. 
+/// 
+///     The output is assumed to be representable by a string, but subclasses
+///     may choose their own runtime representation.
+/// 
+///     The class indicates which formats are supported via a list of string identifiers.
+///     These identifiers can be used to register the class with a GraphIORegistry.
+/// 
+///     The default traversal logic will call into a set of utities
+///     which are responsible for emitting the appropriate output in the desired format.
+///     A derived class may chose to implement their own traversal logic as well.
+///     
+/// </summary>
+class MX_FORMAT_API GraphIO
+{
+  public:
+    GraphIO(){};
+    virtual ~GraphIO(){};
+
+    /// Returns a list of formats that the GraphIO can convert from to
+    const StringSet& supportsFormats() const
+    {
+        return _formats;
+    }
+
+    /// @name I/O methods
+    /// @{
+
+    /// Traverse a graph and return a string withe formatted output
+    /// Derived classes must implement this method.
+    /// @param graph GraphElement to write
+    /// @param roots Optional list of roots to GraphIO what upstream elements to consider>
+    /// @returns String result
+    virtual string write(GraphElementPtr graph, const std::vector<OutputPtr> roots) = 0;
+
+    /// @}
+    /// @name Options to set before writing
+    /// @{
+
+    /// Get options for generation
+    const GraphIOGenOptions& getGenOptions() const
+    {
+        return _genOptions;
+    }
+
+    /// Set options for generation
+    void setGenOptions(const GraphIOGenOptions& options)
+    {
+        _genOptions = options;
+    }
+
+  protected:
+    /// @name Generation Utilties
+    /// @{
+
+    /// Traverse a graph and call into utility methods to emit which handle emitting the 
+    /// appropriate output. 
+    /// @param graph GraphElement to traverse
+    /// @param roots Optional list of roots indicating what upstream elements to consider.
+    ///              If empty, then all of the graph's outputs are traversed.
+    virtual void emitGraph(GraphElementPtr graph, const std::vector<OutputPtr> roots);
```

Comment:
Reduce complexity by only requiring derived classes to support emitting nodes and edges.